osu.Game.Tests/Visual/UserInterface/TestSceneModPresetColumn.cs

```diff
@@ -5,40 +5,148 @@
 using System.Linq;
 using NUnit.Framework;
 using osu.Framework.Allocation;
+using osu.Framework.Extensions.ObjectExtensions;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
+using osu.Framework.Testing;
 using osu.Game.Overlays;
 using osu.Game.Overlays.Mods;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mania.Mods;
 using osu.Game.Rulesets.Mods;
 using osu.Game.Rulesets.Osu.Mods;
 
 namespace osu.Game.Tests.Visual.UserInterface
 {
     public class TestSceneModPresetColumn : OsuTestScene
     {
+        protected override bool UseFreshStoragePerRun => true;
+
+        private RulesetStore rulesets = null!;
+
         [Cached]
         private OverlayColourProvider colourProvider = new OverlayColourProvider(OverlayColourScheme.Green);
 
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Dependencies.Cache(rulesets = new RealmRulesetStore(Realm));
+            Dependencies.Cache(Realm);
+        }
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddStep("clear contents", Clear);
+            AddStep("reset storage", () =>
+            {
+                Realm.Write(realm =>
+                {
+                    realm.RemoveAll<ModPreset>();
+
+                    var testPresets = createTestPresets();
+                    foreach (var preset in testPresets)
+                        preset.Ruleset = realm.Find<RulesetInfo>(preset.Ruleset.ShortName);
```

Comment:
This refetch is a little annoying, but unfortunately required, as far as I can tell. The issue here is that `RulesetStore` is providing detached model instances, therefore confusing realm into trying to insert them again and causing primary key conflicts as a result. Notably, [similar things are already being done in other places](https://github.com/ppy/osu/blob/998e3b74d60180ddfeb4ee63b25b886e4f3db9f7/osu.Game/Database/RealmObjectExtensions.cs#L44).

Reply:
It does beg the question of whether we should even be linking the two entities in the first place (are we gaining anything from it? probably no) rather than just storing the ruleset's `ShortName`. Something I've been thinking about each time this comes up, no need for immediate action or anything.